## 错误处理
 

### 代码实例

```
// Go 语言使用一个独立的·明确的返回值来传递错误信息的。这与
// 使用异常的 Java 和 Ruby 以及在 C 语言中经常见到的超重的
// 单返回值/错误值相比，Go 语言的处理方式能清楚的知道哪个函数
// 返回了错误，并能想调用那些没有出错的函数一样调用。

package main

import &#34;errors&#34;
import &#34;fmt&#34;

// 按照惯例，错误通常是最后一个返回值并且是 `error` 类
// 型，一个内建的接口。
func f1(arg int) (int, error) {
    if arg == 42 {

        // `errors.New` 构造一个使用给定的错误信息的基本
        // `error` 值。
        return -1, errors.New(&#34;can&#39;t work with 42&#34;)

    }

    // 返回错误值为 nil 代表没有错误。
    return arg + 3, nil
}

// 通过实现 `Error` 方法来自定义 `error` 类型是可以得。
// 这里使用自定义错误类型来表示上面的参数错误。
type argError struct {
    arg  int
    prob string
}

func (e *argError) Error() string {
    return fmt.Sprintf(&#34;%d - %s&#34;, e.arg, e.prob)
}

func f2(arg int) (int, error) {
    if arg == 42 {

        // 在这个例子中，我们使用 `&amp;argError` 语法来建立一个
        // 新的结构体，并提供了 `arg` 和 `prob` 这个两个字段
        // 的值。
        return -1, &amp;argError{arg, &#34;can&#39;t work with it&#34;}
    }
    return arg + 3, nil
}

func main() {

    // 下面的两个循环测试了各个返回错误的函数。注意在 `if`
    // 行内的错误检查代码，在 Go 中是一个普遍的用法。
    for _, i := range []int{7, 42} {
        if r, e := f1(i); e != nil {
            fmt.Println(&#34;f1 failed:&#34;, e)
        } else {
            fmt.Println(&#34;f1 worked:&#34;, r)
        }
    }
    for _, i := range []int{7, 42} {
        if r, e := f2(i); e != nil {
            fmt.Println(&#34;f2 failed:&#34;, e)
        } else {
            fmt.Println(&#34;f2 worked:&#34;, r)
        }
    }

    // 你如果想在程序中使用一个自定义错误类型中的数据，你
    // 需要通过类型断言来得到这个错误类型的实例。
    _, e := f2(42)
    if ae, ok := e.(*argError); ok {
        fmt.Println(ae.arg)
        fmt.Println(ae.prob)
    }
}
```

### 运行程序

```
$ go run errors.go
f1 worked: 10
f1 failed: can&#39;t work with 42
f2 worked: 10
f2 failed: 42 - can&#39;t work with it
42
can&#39;t work with it

# 到 Go 博客去看这篇[很棒的文章](http://blog.golang.org/2011/07/error-handling-and-go.html)
# 获取更多关于错误处理的信息。
```

## 协程
 

### 代码实例

```
// _Go 协程_ 在执行上来说是轻量级的线程。

package main

import &#34;fmt&#34;

func f(from string) {
    for i := 0; i &lt; 3; i++ {
        fmt.Println(from, &#34;:&#34;, i)
    }
}

func main() {

    // 假设我们有一个函数叫做 `f(s)`。我们使用一般的方式
    // 调并同时运行。
    f(&#34;direct&#34;)

    // 使用 `go f(s)` 在一个 Go 协程中调用这个函数。
    // 这个新的 Go 协程将会并行的执行这个函数调用。
    go f(&#34;goroutine&#34;)

    // 你也可以为匿名函数启动一个 Go 协程。
    go func(msg string) {
        fmt.Println(msg)
    }(&#34;going&#34;)

    // 现在这两个 Go 协程在独立的 Go 协程中异步的运行，所以
    // 我们需要等它们执行结束。这里的 `Scanln` 代码需要我们
    // 在程序退出前按下任意键结束。
    var input string
    fmt.Scanln(&amp;input)
    fmt.Println(&#34;done&#34;)
}
```

### 运行程序

```
# 当我们运行这个程序时，将首先看到阻塞式调用的输出，然后是
# 两个 Go 协程的交替输出。这种交替的情况表示 Go 运行时是以
# 异步的方式运行协程的。
$ go run goroutines.go
direct : 0
direct : 1
direct : 2
goroutine : 0
going
goroutine : 1
goroutine : 2
&lt;enter&gt;
done

# 接下来我们将学习在并发的 Go 程序中的 Go 协程的辅助
# 特性：通道。
```